cezo_fl/client.py

```diff
@@ -0,0 +1,138 @@
+                device=self.device,
+            )
+        else:
+            raise Exception(f'{grad_estimator_params["method"]} is not supported')
+
+    def _initialize_optimizer(self, optimzier_params):
+        if optimzier_params["method"] == "SGD":
+            return SGD(
+                self.model.parameters(),
+                lr=optimzier_params["lr"],
+                momentum=optimzier_params["momentum"],
+                weight_decay=1e-5,
+            )
+        else:
+            raise Exception(f'{optimzier_params["method"]} is not supported')
+
+    def _get_train_batch_iterator(self):
+        # NOTE: used only in init, will generate an infinite iterator from dataloader
+        while True:
+            for v in self.dataloader:
+                yield v
+
+    def local_update(self, seeds: Sequence[int]) -> Sequence[torch.Tensor]:
+        """Returns a sequence of gradient scalar tensors for each local update.
+
+        The length of the returned sequence should be the same as the length of seeds.
+        The inner tensor can be a scalar or a vector. The length of vector is the number
+        of perturbations.
+        """
+        ret: Sequence[torch.Tensor] = []
+        for seed in seeds:
+            # NOTE:dataloader manage its own randomnes state thus not affected by seed
+            batch_inputs, labels = next(self.data_iterator)
+            if self.device != torch.device("cpu"):
+                batch_inputs, labels = batch_inputs.to(self.device), labels.to(
+                    self.device
+                )
+            # generate grads and update model's gradient
+            torch.manual_seed(seed)
+            seed_grads = self.grad_estimator.compute_grad(
+                batch_inputs, labels, self.criterion
+            )
+            ret.append(seed_grads)
+
+            # update model
+            # NOTE: local model update also uses momentum and other states
+            self.optimizer.step()
+
+        return ret
+
+    def reset_model(self) -> None:
+        """Reset the mode to the state before the local_update."""
+        self.model.load_state_dict(self.last_pull_state_dict["model"])
+        self.optimizer.load_state_dict(self.last_pull_state_dict["optimizer"])
+
+    def screenshot(self) -> dict:
+        # deepcopy current model.state_dict and optimizer.state_dict
+        return deepcopy(
+            {"model": self.model.state_dict(), "optimizer": self.optimizer.state_dict()}
+        )
+
+    def pull_model(
+        self,
+        seeds_list: Sequence[Sequence[int]],
+        gradient_scalar: Sequence[Sequence[torch.Tensor]],
+    ) -> None:
+        # reset model
+        self.reset_model()
+        # update model to latest version
+        for iteration_seeds, iteration_grad_avgs in zip(seeds_list, gradient_scalar):
+            for local_update_seed, local_update_grad_vector in zip(
+                iteration_seeds, iteration_grad_avgs
+            ):
+                # create gradient
```

Comment:
I think you missed `self.optimizer.zero_grad()`.

Also you can merge this with `update_model_given_seed_and_grad` in server.py into one shared function?